modules/guides/src/GuidesStorage.php

```diff
@@ -0,0 +1,155 @@
+
+  /**
+   * The settings of the site.
+   *
+   * @var \Drupal\Core\Site\Settings
+   */
+  protected $settings;
+
+  /**
+   * The cache backend service.
+   *
+   * @var \Drupal\Core\Cache\CacheBackendInterface
+   */
+  protected $cacheBackend;
+
+  /**
+   * GuidesStorage constructor.
+   *
+   * @param \Drupal\Core\Site\Settings $settings
+   *   The settings of the site.
+   * @param \Drupal\Core\Cache\CacheBackendInterface $cache_backend
+   *   The cache backend.
+   */
+  public function __construct(Settings $settings, CacheBackendInterface $cache_backend) {
+    $this->settings = $settings;
+    $this->cacheBackend = $cache_backend;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  private function getDirectory(): string {
+    return DRUPAL_ROOT . ($this->settings->get('guides_dir') ?? '/guides');
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getFiles(): array {
+    $directory = $this->getDirectory();
+    $cid = 'guides_files:' . $directory;
+    $data_cached = $this->cacheBackend->get($cid);
+
+    if ($data_cached) {
+      $guides = $data_cached->data;
+    }
+    else {
+      $iterator = new \RecursiveDirectoryIterator($directory, \FilesystemIterator::CURRENT_AS_FILEINFO | \FilesystemIterator::SKIP_DOTS);
+      $filter = new \RecursiveCallbackFilterIterator($iterator, function ($current, $key, $iterator) {
+        if ($iterator->hasChildren()) {
+          return TRUE;
+        }
+        $path = $current->getPathname();
+        if (is_file($path)) {
+          return substr($current->getFilename(), -3) == '.md';
+        }
+
+        return FALSE;
+      });
+
+      $guides = [];
+      $files = new \RecursiveIteratorIterator($filter);
+      foreach ($files as $md) {
+        if ($files->getDepth() == 1) {
+          $guides[] = $md->getPathname();
+        }
+      }
+
+      $this->cacheBackend->set($cid, $guides);
+    }
+
+    return $guides;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getFile(string $subdirectory, string $file_name): string {
+    $file_path = $this->getDirectory() . '/' . $subdirectory . '/' . $file_name . '.md';
+
+    if (file_exists($file_path) && in_array($file_path, $this->getFiles())) {
+      return $file_path;
+    }
+    else {
+      throw new FileNotFoundException($file_path);
+    }
+
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getFileContent(string $subdirectory, string $file_name): string {
+    $file_path = $this->getFile($subdirectory, $file_name);
+    if (file_exists($file_path)) {
+      $guide_directory = dirname($file_path);
+      $content = str_replace('@guide_path', $guide_directory, file_get_contents($file_path));
+      return $content;
+    }
+    else {
+      throw new FileNotFoundException($file_path);
+    }
+
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getLinks(): array {
+    $guides_files = $this->getFiles();
+    $guides = [];
+
+    foreach ($guides_files as $md) {
+      if (file_exists($md)) {
```

Comment:
Let's not check whether the file exists here because if we have 50 files it means 50 file system operation which is not so much better than loading a guides directory without caching for every request.
Let's document in this methods doc block that this method intentionally does not validate whether a file _still_ exists or not, but the `getFile()` method validates it and because it is called by the route access handler, therefore, the `http://example.com/guides/file_does_not_exists_anymore` page will return a 404.
In other words, the worst-case we render some links that are going to be 404, but we only check whether a file _still_ exists or not when someone clicks on a link.